### PR TITLE
Ensure accelerator is valid if running interactively

### LINKED ISSUE #5966: Add ipython kernel detection and give warning when accelerator = "ddp"
## 🚀 Feature

Add ipython kernel detection and give warning when accelerator = "ddp"
### Motivation
When users try to use `ddp` as accelerator in Jupyter Notebook or Jupyter Lab, the trainer will be stuck forever and no hints about the cause. So, to better inform developers, it will be great to detect whether the code is run in a ipython kernel.

### Pitch

When users initialize a trainer by `Trainer(accelerator = 'ddp')` in a ipython kernel, give a warning or exception. Something like "ddp will not work in ipython kernel" should suffice.

### Additional context
Please see [this](https://github.com/ipython/ipython/issues/11694#issuecomment-485802013) to implement it. I can do a PR, but I am not sure if it is appropriate to place code for detection in `Trainer.__init__()`

### PR DESCRIPTION
**Pull request by ifsheldon**
## What does this PR do?

Fixes #5966 
Raise an exception rather than a warning to force users to use compatible accelerators.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/accelerator_connector.py [hunk 1]
```diff
@@ -500,7 +500,28 @@ def set_distributed_mode(self, distributed_backend: Optional[str] = None):
             else:
                 rank_zero_warn('You are running on single node with no parallelization, so distributed has no effect.')
                 self._distrib_type = None
-
+        
+        # finished configuring self._distrib_type, check ipython environment
+        ipython_incompatible_distrib_types = [DistributedType.DDP, DistributedType.DDP2, DistributedType.HOROVOD]
```
Comment (tchaton) [inline pytorch_lightning/accelerators/accelerator_connector.py:505]: It should be simpler to list the compatible ones as only `DP`, `DDP_SPAWN ` and `DDP_SHARDED_SPAWN ` should be.

```
class DistributedType(LightningEnum):
    """ Define type of ditributed computing.

    >>> # you can math the type with string
    >>> DistributedType.DDP == 'ddp'
    True
    >>> # which is case invariant
    >>> DistributedType.DDP2 in ('ddp2', )
    True
    """
    DP = 'dp'
    DDP = 'ddp'
    DDP2 = 'ddp2'
    DDP_SPAWN = 'ddp_spawn'
    HOROVOD = 'horovod'
    DDP_SHARDED = 'ddp_sharded'
    DDP_SHARDED_SPAWN = 'ddp_sharded_spawn'
    RPC_SEQUENTIAL_PLUGIN = 'rpc_sequential'
```

File: pytorch_lightning/accelerators/accelerator_connector.py [hunk 2]
```diff
@@ -500,7 +500,28 @@ def set_distributed_mode(self, distributed_backend: Optional[str] = None):
...
+        if self._distrib_type in ipython_incompatible_distrib_types:
+            # check ipython env
```
Comment (tchaton) [inline pytorch_lightning/accelerators/accelerator_connector.py:507]: Could you put this logic into its own function ? Like `_check_ipython_compatible`

**Review by tchaton — COMMENTED** [review]
Overall, looks good. Some nits to resolve. Thanks for your work !

**Review by tchaton — COMMENTED** [review]
Overall looks good ! Small nits left !

File: pytorch_lightning/accelerators/accelerator_connector.py (hunk 1 above)
Comment (ifsheldon) [reply pytorch_lightning/accelerators/accelerator_connector.py]: Make sense, so I added an attribute `self.ipython_compatible_distrib_types` to record compatibility in case of future compatibility improvements

File: pytorch_lightning/accelerators/accelerator_connector.py (hunk 2 above)
Comment (ifsheldon) [reply pytorch_lightning/accelerators/accelerator_connector.py]: Done, not sure the naming fits the convention, but I see `check_horovod()` is okay.

File: pytorch_lightning/accelerators/accelerator_connector.py [hunk 3]
```diff
@@ -542,6 +546,27 @@ def _set_horovod_backend(self):
         else:
             self.num_processes = hvd.local_size()
 
+    def check_ipython_compatibility(self):
+        """Raises a `MisconfigurationException` if the accelerator is not compatible with IPython and code is run in an IPython kernel."""
+        if self._distrib_type not in self.ipython_compatible_distrib_types:
+            # check ipython env
+            import sys
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/accelerator_connector.py:553]: can you put the import at the top level?

File: pytorch_lightning/accelerators/accelerator_connector.py [hunk 4]
```diff
@@ -542,6 +546,27 @@ def _set_horovod_backend(self):
...
+                in_ipython = ip is not None
+
+            if in_ipython:
+                in_ipython_kernel = getattr(ip, 'kernel', None) is not None
+            
+            if in_ipython_kernel:
+                raise MisconfigurationException(
+            "Selected distributed backend {self._distrib_type} not compatible with IPython environment"
+            "Run your code as a script, or choose one of compatible backends {self.ipython_compatible_distrib_types} as accelerator backend"
+            )
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/accelerator_connector.py:568]: needs to be f string otherwise this won't work
also could you run `yapf -i` on this file to get proper formatting?

File: pytorch_lightning/accelerators/accelerator_connector.py [hunk 5]
```diff
@@ -542,6 +546,27 @@ def _set_horovod_backend(self):
         else:
             self.num_processes = hvd.local_size()
 
+    def check_ipython_compatibility(self):
+        """Raises a `MisconfigurationException` if the accelerator is not compatible with IPython and code is run in an IPython kernel."""
+        if self._distrib_type not in self.ipython_compatible_distrib_types:
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/accelerator_connector.py:551]: you can invert the if statement and return early.

**Review by awaelchli — COMMENTED** [review]
Could you comment on the robustness of this approach?
Can there be instances where IPython is loaded but we're not in a notebook? We should make sure we don't raise any false positive error messages

**Comment by ifsheldon** [comment]
> Could you comment on the robustness of this approach?
> Can there be instances where IPython is loaded but we're not in a notebook? We should make sure we don't raise any false positive error messages

I would say it is reliable in my several tests on Linux, Windows and macOS with different distributions of IPython(x64 and aarch). And the code is [given](https://github.com/ipython/ipython/issues/11694#issuecomment-485802013) by the maintainers of IPython, which, I think, add more reliability. 
And I dinged into the code a bit. The code calls [get_ipython()](https://github.com/ipython/ipython/blob/master/IPython/core/getipython.py#L17) that returns a `InteractiveShell` instance. That seems to be indeed reasonable way to check IPython environment, although I cannot find when the attribute `kernel` is attached to a `InteractiveShell` instance. Maybe @minrk can be so kind as to comment on that.

File: pytorch_lightning/accelerators/accelerator_connector.py (hunk 5 above)
Comment (ifsheldon) [reply pytorch_lightning/accelerators/accelerator_connector.py]: Done

File: pytorch_lightning/accelerators/accelerator_connector.py (hunk 4 above)
Comment (ifsheldon) [reply pytorch_lightning/accelerators/accelerator_connector.py]: Done and sorry for the mistake

File: pytorch_lightning/accelerators/accelerator_connector.py (hunk 3 above)
Comment (ifsheldon) [reply pytorch_lightning/accelerators/accelerator_connector.py:557]: Since here is the only place where `sys` is used, I am hesitating on whether to move it up to the top level, but if you insist, it is fine for me.

**Comment by ifsheldon** [comment]
Can you please detail which kinds of `IPython` kernel are not compatible? I just found out that Jupiter-lab and Jupyter-notebook use `zmqkernel`, which means `get_ipython()` will return an instance of `ipykernel.zmqshell.ZMQInteractiveShell` that has a `kernel` attribute. However, if I run `get_ipython()` in terminal inside `ipython`, I got an instance of `IPython.terminal.interactiveshell.TerminalInteractiveShell` that has no `kernel` attribute, and `ddp` works well with that interactive shell, testing on the boring model with 2 GPUs.

**Comment by ifsheldon** [comment]
`get_ipython()` explicitly check IPython environment. If python codes is run in IPython, it will return a shell instance. And the following `getattr(ip, 'kernel', None) is not None` only detects whether a kernel is present and implicitly infer whether python code is run with Jupyter, since Jupyter uses zmq interactive shell that has `kernel` attribute.

So, if we want to just check IPython, `get_ipython is not None` should suffice. But, if we want to detect Jupyter frontend, there is no way(as IPython aims to separate frontends and backends) but to infer that from shell types.

**Comment by awaelchli** [comment]
> Can you please detail which kinds of IPython kernels and shells are not compatible? Or can you be more specific on what went wrong with Jupyter-lab and Jupyter notebook that causes such incompatibility?

DDP incompatibility is not strictly just with "notebooks". Our observation is simply that it doesn't work with them. 

A script that contains `Trainer(accelerator=ddp, gpus=4)` must be launched with a python command, i.e.
python train.py.
Under the hood, the process (call it main process) will then launch other processes (in the example here 3) using subprocess.Popen, so essentially:

```
LOCAL_RANK=1 {... other env vars} python train.py
LOCAL_RANK=2 {... other env vars} python train.py
LOCAL_RANK=3 {... other env vars} python train.py
```

This can't work with notebooks, because notebooks can't be launched this way. Same with interactive shell I guess...

**Comment by ifsheldon** [comment]
I see, and I checked again and I saw some weird issue when I run the boring model with `ddp` in IPython interactive shell. So, then only checking ipython should suffice, and I modified my code as the new commit. This should work as long as `get_ipython()` works.

File: pytorch_lightning/accelerators/accelerator_connector.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator_connector.py:557]: In Lightning it's always at the top no matter how many times it is used
cc @Borda

Comment (Borda) [reply pytorch_lightning/accelerators/accelerator_connector.py]: yes, agree...

Comment (ifsheldon) [reply pytorch_lightning/accelerators/accelerator_connector.py]: OK, moved it

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 6]
```diff
@@ -84,6 +85,9 @@ def __init__(
         # initialization
         self._device_type = DeviceType.CPU
         self._distrib_type = None
+        self.ipython_compatible_distrib_types = [
+            DistributedType.DP, DistributedType.DDP_SPAWN, DistributedType.DDP_SHARDED_SPAWN
+        ]
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:90]: maybe it would be better to move this into the `DistributedType` class as a method that returns the list of incompatible types. 
@carmocca suggested something similar in another place.

**Review by awaelchli — APPROVED** [review]
@ifsheldon I fixed the pep8 problems (line too long) and added a test for coverage, hope this is fine with you.
otherwise LGTM

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 6 above)
Comment (ifsheldon) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:90]: Thanks for the test! Maybe it would be better to add some tests on the compatible ones to test false negative? The suggestion makes sense. I thought it would be better to make ipython compatibility an attribute of this enum, but a method returning a list seems good enough for now

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py]: Created #6079 to track this

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 7]
```diff
@@ -556,6 +560,23 @@ def _set_horovod_backend(self):
         else:
             self.num_processes = hvd.local_size()
 
+    def check_ipython_compatibility(self):
+        """
+        Raises a `MisconfigurationException` if the accelerator and/or plugin is not compatible with IPython
+        and code is run in an IPython kernel.
+        """
+        if self._distrib_type in DistributedType.ipython_compatible_types():
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:568]: This should be

```suggestion
        if self._distrib_type.is_ipython_compatible():
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 8]
```diff
@@ -556,6 +560,23 @@ def _set_horovod_backend(self):
...
+            return
+        # check ipython env
+        if "IPython" in sys.modules:
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:571]: Use https://github.com/PyTorchLightning/pytorch-lightning/blob/049006a59cf62d3539aaf19037aca1caac98abfd/pytorch_lightning/utilities/imports.py#L24

Comment (ifsheldon) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py]: Done

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 7 above)
Comment (ifsheldon) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py]: Make sense. Done

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 9]
```diff
@@ -556,6 +560,23 @@ def _set_horovod_backend(self):
...
+        """
+        Raises a `MisconfigurationException` if the accelerator and/or plugin is not compatible with IPython
+        and code is run in an IPython kernel.
+        """
+        if self._distrib_type is None or self._distrib_type.is_ipython_compatible():
+            return
+        # check ipython env
+        if _module_available("IPython"):
+            from IPython import get_ipython
+            if get_ipython() is not None:
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:573]: noob question: when would this be None?

Comment (ifsheldon) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:573]: `get_ipython()` will return `None` when the code is not run in a ipython environment(like notebook or ipython interactive shell)

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 10]
```diff
@@ -568,7 +568,7 @@ def check_ipython_compatibility(self):
         if self._distrib_type in DistributedType.ipython_compatible_types():
             return
         # check ipython env
-        if "IPython" in sys.modules:
+        if _module_available("IPython"):
```
Comment (ifsheldon) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:571]: The failing tests seem to be related to this line. Any ideas?

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:571]: Any idea why `IPython` does not define `__spec__` @Borda?

Should we keep `"IPython" in sys.modules` or update `_module_available`?

Comment (ifsheldon) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:571]: Any suggestions? I think if we keep `"IPython" in sys.modules`, this PR is ready to be merged.

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:571]: Fixed it. You forgot to remove

`@mock.patch.dict(sys.modules, {"IPython": Mock()})`

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 10 above)
Comment (ifsheldon) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:573]: Thanks for the patch and sorry for my sloppiness. But I see even after your commit, there are still failing test cases on `test_accelerator_connector.py::test_ipython_incompatible_backend_error` and there's not so much useful information provided in the log. Are the failures related to `_module_available()` or to `get_ipython()`?

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:573]: Maybe we should use this to check if it's interactive

https://stackoverflow.com/a/64523765

Instead of relying on IPython

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py]: Done in https://github.com/PyTorchLightning/pytorch-lightning/pull/5970/commits/990c03ef6ecafb4c3a82a85bb54ac89c6d3bfcb4

**Review by s-rog — APPROVED** [review]

File: tests/accelerators/test_accelerator_connector.py [hunk 11]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License
 
 import os
+import sys
```
Comment (carmocca) [inline tests/accelerators/test_accelerator_connector.py:16]:
```suggestion
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 10 above)
Comment (ifsheldon) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py]: Seems reasonable. Thanks!

**Comment by awaelchli** [comment]
Thanks for sending in the PR @ifsheldon and also @carmocca for the help polishing it
